### PR TITLE
operator: expose build-image API

### LINKED ISSUE #304: build-image API: expose the APIs
* add a seed image url field (in MachineRegistration)
* expose the BUILD API in the operator
* do a PoC build inside the operator (quick and dirty)
* expose the STATUS API

### PR DESCRIPTION
**Pull request by fgiudici**
This PR add better parsing of the URL of the incoming http connections and split the code belonging to different APIs in different files.
The only available API we had was the _registration_ one.
This PR adds the new _build-iso_ API, which supports POST and GET endpoints to respectively start and get the current status of build image jobs.
Note that this PR just exposes the API, but the actual creation of the image is not implemented: so, the code will always fail the build at present.

This is the first step in order to address #207 and IMO is enough to consider #304 complete.

While it is obviously not complete (there is no implementation of the build job) it could be merged without harm: the new API will be exposed and if called will just report the build to have failed.
We could also wait to merge to have a basic build functionality: in this case I would ask to hold any change affect the registration code that may not be urgent, as this PR does a reorganization of the functions in different files and rebasing may be a pain 😅 

Short description of the API can be found on the [wiki](https://github.com/rancher/elemental-operator/wiki/Elemental-Opreator-extended-APIs#build-image-api).

Fixes #304

### CODE REVIEW

File: pkg/server/api_buildimage.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+			return errMsg
+		}
+	}
+
+	i.registrationCache.setRegistrationData(job.Token, registrationData{
+		buildImageUrl:    job.Url,
+		buildImageStatus: jobStatusStarted,
+	})
+
+	logrus.Infof("New build-image job queued: seed image:'%s', reg token:'%s'", job.Url, job.Token)
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go]: Duplicated

File: pkg/server/api_buildimage.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		http.Error(resp, err.Error(), http.StatusBadRequest)
+		return err
+	}
+
+	if reg, err := i.registrationCache.getRegistrationData(job.Token); err != nil {
+		if reg.buildImageStatus == jobStatusStarted {
+			if reg.buildImageUrl == job.Url {
+				logrus.Debugf("Same job already started, nothing to do")
+
+				if err := json.NewEncoder(resp).Encode(job); err != nil {
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:113]: job members are sanitized at lines 100-101

File: pkg/server/api_buildimage.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+func (i *InventoryServer) draftDoBuildImage(job buildImageJob) {
+	// This should spawn a pod that will:
+	// - download the base image (job.Url)
+	// - retrieve the registration YAML from the registrationToken (job.Token)
+	// - generate the seed image (base image + registration yaml)
+	// - return the seed image (url? shared volume?)
+	// Here we should just update the status then.
+	// For now, it will always fail.
+	logrus.Infof("build-image for token %s failed (build job not supported yet).", job.Token)
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:151]: job members are sanitized at lines 100-101

File: pkg/server/server.go [hunk 4]
```diff
@@ -48,37 +54,76 @@
...
+			return
+		}
+	case "registration":
+		if err := i.apiRegistration(resp, req); err != nil {
+			logrus.Errorf("registration: %s", err.Error())
+			return
 		}
+	default:
+		logrus.Errorf("Unknown API: %s", api)
+		http.Error(resp, fmt.Sprintf("unknwon api: %s", api), http.StatusBadRequest)
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/server.go:95]: Fixed, as URL is sanitized at line 68 before splitting it in parts (from which we derive _api_ )

**Comment by fgiudici** [comment]
All github code scanning complaints have been addressed (I dismissed one by one with comments but there is no way to let them stay folded).

File: pkg/server/api_buildimage.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:151]: job is sanitized at lines 100-101

File: pkg/server/api_buildimage.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	if err := i.registrationCache.setBuildImageStatus(job.Token, jobStatusFailed); err != nil {
+		logrus.Errorf("Cannot update build-image status for job with token %s: %s", job.Token, err.Error())
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:153]: job sanitized at lines 100-101

File: pkg/server/api_buildimage.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+			return errMsg
+		}
+	}
+
+	i.registrationCache.setRegistrationData(job.Token, registrationData{
+		buildImageURL:    job.URL,
+		buildImageStatus: jobStatusStarted,
+	})
+
+	logrus.Infof("New build-image job queued: seed image:'%s', reg token:'%s'", job.URL, job.Token)
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:131]: job sanitized at lines 100-101

[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:131]: job members sanitized at lines 100-101.
This is also duplicated ! 🤦🏼

File: pkg/server/api_buildimage.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	i.registrationCache.setRegistrationData(job.Token, registrationData{
+		buildImageUrl:    job.Url,
+		buildImageStatus: jobStatusStarted,
+	})
+
+	logrus.Infof("New build-image job queued: seed image:'%s', reg token:'%s'", job.Url, job.Token)
+	// start the actual build job here
+	go i.draftDoBuildImage(job)
+
+	if err := json.NewEncoder(resp).Encode(job); err != nil {
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/server/api_buildimage.go:135]: job sanitized at lines 100-101

File: pkg/server/server.go [hunk 8]
```diff
@@ -48,37 +55,77 @@ func New(ctx context.Context, cl client.Client) *InventoryServer {
...
+		CheckOrigin:      func(r *http.Request) bool { return true },
 	}
 
-	if setting.Value == "" {
-		return "", fmt.Errorf("server-url is not set")
+	conn, err := upgrader.Upgrade(resp, req, nil)
+	if err != nil {
+		return nil, err
 	}
+	_ = conn.SetWriteDeadline(time.Now().Add(register.RegistrationDeadlineSeconds * time.Second))
```
Comment (alexander-demicev) [inline pkg/server/server.go:117]: should we check for error here?

File: pkg/server/api_buildimage_test.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	"testing"
+
+	elementalv1 "github.com/rancher/elemental-operator/api/v1beta1"
+	"gotest.tools/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestApiBuildImage(t *testing.T) {
```
Comment (alexander-demicev) [inline pkg/server/api_buildimage_test.go:33]: what do you think about using ginkgo/gomega here? this way stay aligned with controllers tests.

Comment (davidcassany) [reply pkg/server/api_buildimage_test.go:33]: Yes I agree, having all tests using the same framework is helpful for maintenance, it shouldn't be a big deal using ginkgo and gomega.

File: pkg/server/api_buildimage.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+	"io"
+	"net/http"
+	"path"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	jobStatusStarted = "Started"
+	//	jobStatusCompleted  = "Completed"
```
Comment (davidcassany) [inline pkg/server/api_buildimage.go:31]: not sure why this is a comment, even I guess it is not relevant at this stage

**Review by davidcassany — APPROVED** [review]
Looks good to me.

From the API definition and docs it is unclear to me where we get the built ISO URL from. Probably I would include it as part of the result of `/build-image/{token} GET` and only provide a URL on `Completed` status. In any case this can be easily adapted and tuned as part of the job implementation.

**Comment by davidcassany** [comment]
> We could also wait to merge to have a basic build functionality: in this case I would ask to hold any change affect the registration code that may not be urgent, as this PR does a reorganization of the functions in different files and rebasing may be a pain sweat_smile

I am more in favor of merging, if we need any urgent backport we can simply fork a maintenance branch, my bet is it won't be needed. Note the new kubebuilder operator is not yet released so urgent changes are only targeting v1.0.x and for this series a maintenance branch is likely to be needed anyways if we get bug reports.

File: pkg/server/server.go (hunk 8 above)
Comment (fgiudici) [reply pkg/server/server.go:117]: *SetWriteDeadline* always return nil error but (the immediately following) *SetReadDeadline* may fail:
agreed, we should report the error but try to serve the websocket in any case.

File: pkg/server/api_buildimage_test.go (hunk 9 above)
Comment (fgiudici) [reply pkg/server/api_buildimage_test.go:33]: agreed

**Comment by fgiudici** [comment]
> Looks good to me.
> 
> From the API definition and docs it is unclear to me where we get the built ISO URL from. Probably I would include it as part of the result of `/build-image/{token} GET` and only provide a URL on `Completed` status. In any case this can be easily adapted and tuned as part of the job implementation.

Oh right, that was the plan I had in mind (also if right now the GET returns just the status)

File: pkg/server/api_buildimage.go (hunk 10 above)
Comment (fgiudici) [reply pkg/server/api_buildimage.go:31]: It was just to make the linter happy (as right now that const is not used 😅 )

File: pkg/server/api_buildimage_test.go (hunk 9 above)
Comment (fgiudici) [reply pkg/server/api_buildimage_test.go:33]: going to merge this, will do soon in a following PR

**Pull request merged**